File: src/fetcher.py
from __future__ import annotations

import json
import subprocess

from models import Issue, Phase, Priority, RepositoryConfig


class GitHubCLIError(Exception):
    """gh コマンドの実行エラー"""

    pass


class IssueParseError(Exception):
    """Issue の JSON パースエラー"""

    pass


def fetch_issues(repo_config: RepositoryConfig, phase: Phase) -> list[Issue]:
    """指定リポジトリ・フェーズの Issue を取得し、優先度順にソートして返す。

    Args:
        repo_config: リポジトリ設定
        phase: 処理フェーズ

    Returns:
        優先度順にソートされた Issue リスト

    Raises:
        GitHubCLIError: gh コマンドの実行に失敗した場合
        IssueParseError: JSON のパースに失敗した場合
    """
    if phase == Phase.PLAN:
        trigger_label = repo_config.labels.plan.trigger
    else:
        trigger_label = repo_config.labels.impl.trigger

    args = [
        "gh",
        "issue",
        "list",
        "--repo",
        repo_config.full_name,
        "--label",
        trigger_label,
        "--state",
        "open",
        "--json",
        "number,title,body,labels,url",
        "--limit",
        "100",
    ]

    raw_json = _run_gh_command(args)
    issues = _parse_issues(raw_json, phase, repo_config)
    return _sort_by_priority(issues)


def _run_gh_command(args: list[str]) -> str:
    """gh コマンドを実行し、stdout を返す。

    Args:
        args: コマンドライン引数のリスト

    Returns:
        コマンドの標準出力

    Raises:
        GitHubCLIError: コマンドの終了コードが 0 でない場合
    """
    result = subprocess.run(args, shell=False, capture_output=True, text=True)

    if result.returncode != 0:
        raise GitHubCLIError(result.stderr)

    return result.stdout


def _parse_issues(
    raw_json: str, phase: Phase, repo_config: RepositoryConfig
) -> list[Issue]:
    """JSON 文字列から Issue リストを生成する。

    Args:
        raw_json: gh コマンドの JSON 出力
        phase: 処理フェーズ
        repo_config: リポジトリ設定

    Returns:
        Issue のリスト

    Raises:
        IssueParseError: JSON のパースに失敗した場合
    """
    try:
        data = json.loads(raw_json)
    except json.JSONDecodeError as e:
        raise IssueParseError(f"Failed to parse JSON: {e}") from e

    issues: list[Issue] = []
    for item in data:
        labels = [label["name"] for label in item.get("labels", [])]
        priority = _determine_priority(labels, repo_config.priority_labels)
        body = item.get("body")

        try:
            issues.append(
                Issue(
                    number=item["number"],
                    title=item["title"],
                    body=body,
                    labels=labels,
                    url=item["url"],
                    phase=phase,
                    priority=priority,
                )
            )
        except KeyError as e:
            raise IssueParseError(f"Missing required field in issue data: {e}") from e

    return issues


def _determine_priority(labels: list[str], priority_labels: list[str]) -> Priority:
    """ラベルから優先度を判定する。

    priority_labels のインデックスで判定:
      - index 0 のラベルが含まれる -> Priority.HIGH
      - index 1 のラベルが含まれる -> Priority.LOW
      - どちらも含まれない -> Priority.NONE

    複数マッチ時は最も優先度が高いもの（インデックスが小さい方）を採用する。

    Args:
        labels: Issue に付与されたラベルのリスト
        priority_labels: 優先度ラベルの定義リスト

    Returns:
        判定された優先度
    """
    priority_map = {
        0: Priority.HIGH,
        1: Priority.LOW,
    }

    for index, priority_label in enumerate(priority_labels):
        if priority_label in labels and index in priority_map:
            return priority_map[index]

    return Priority.NONE


def _sort_by_priority(issues: list[Issue]) -> list[Issue]:
    """Issue を優先度順にソートする。

    ソートキー: (priority.value, number) の昇順

    Args:
        issues: ソート対象の Issue リスト

    Returns:
        ソート済みの Issue リスト
    """
    return sorted(issues, key=lambda i: (i.priority.value, i.number))

File: src/worker/fetcher.ts
import type { Issue, RepositoryConfig } from "./models.js";
import { Phase, Priority, repoFullName } from "./models.js";
import {
  runCommand,
  ProcessTimeoutError,
  ProcessExecutionError,
} from "./process.js";
import { createLogger } from "./logger.js";

/** gh コマンドの実行エラー */
export class GitHubCLIError extends Error {
  constructor(message: string) {
    super(message);
    Object.setPrototypeOf(this, GitHubCLIError.prototype);
  }
}

/** Issue の JSON パースエラー */
export class IssueParseError extends Error {
  constructor(message: string) {
    super(message);
    Object.setPrototypeOf(this, IssueParseError.prototype);
  }
}

const GH_TIMEOUT_MS = 120_000;

const logger = createLogger("fetcher");

/** 処理を許可する authorAssociation の一覧 */
const PERMITTED_ASSOCIATIONS: ReadonlySet<string> = new Set([
  "OWNER",
  "MEMBER",
  "COLLABORATOR",
]);

/** GitHub REST API の Issue レスポンス内の label 構造 */
interface GhLabel {
  readonly name: string;
}

/** GitHub REST API の Issue レスポンス内の 1 Issue 構造 */
interface GhIssueItem {
  readonly number?: number;
  readonly title?: string;
  readonly body?: string | null;
  readonly labels?: readonly GhLabel[];
  readonly html_url?: string;
  readonly author_association?: string;
}

/**
 * 指定リポジトリ・フェーズの Issue を取得し、優先度順にソートして返す。
 *
 * @throws {GitHubCLIError} gh コマンドの実行に失敗した場合
 * @throws {IssueParseError} JSON のパースに失敗した場合
 */
export async function fetchIssues(
  repoConfig: RepositoryConfig,
  phase: Phase,
): Promise<readonly Issue[]> {
  const triggerLabel =
    phase === Phase.PLAN
      ? repoConfig.labels.plan.trigger
      : repoConfig.labels.impl.trigger;

  const args = [
    "api",
    `repos/${repoFullName(repoConfig)}/issues`,
    "--method",
    "GET",
    "--field",
    `labels=${triggerLabel}`,
    "--field",
    "state=open",
    "--field",
    "per_page=100",
  ];

  const rawJson = await runGhCommand(args);
  const issues = parseIssues(rawJson, phase, repoConfig);
  const permitted = filterByAuthorAssociation(issues);
  return sortByPriority(permitted);
}

/**
 * gh コマンドを実行し、stdout を返す。
 *
 * @throws {GitHubCLIError} コマンドの終了コードが 0 でない場合、またはタイムアウト時
 */
async function runGhCommand(args: readonly string[]): Promise<string> {
  try {
    const result = await runCommand("gh", args, { timeoutMs: GH_TIMEOUT_MS });

    if (!result.success) {
      throw new GitHubCLIError(result.stderr);
    }

    return result.stdout;
  } catch (error: unknown) {
    if (error instanceof GitHubCLIError) {
      throw error;
    }
    if (error instanceof ProcessTimeoutError) {
      throw new GitHubCLIError(
        `gh command timed out after ${GH_TIMEOUT_MS / 1_000} seconds`,
      );
    }
    if (error instanceof ProcessExecutionError) {
      throw new GitHubCLIError(error.message);
    }
    throw error;
  }
}

/**
 * JSON 文字列から Issue リストを生成する。
 *
 * @throws {IssueParseError} JSON のパースに失敗した場合
 */
function parseIssues(
  rawJson: string,
  phase: Phase,
  repoConfig: RepositoryConfig,
): Issue[] {
  let data: GhIssueItem[];
  try {
    data = JSON.parse(rawJson) as GhIssueItem[];
  } catch {
    throw new IssueParseError(`Failed to parse JSON: ${rawJson.slice(0, 200)}`);
  }

  return data.map((item) => {
    const labels = (item.labels ?? []).map((label) => label.name);
    const priority = determinePriority(labels, repoConfig.priorityLabels);

    if (item.number === undefined) {
      throw new IssueParseError("Missing required field in issue data: number");
    }
    if (item.title === undefined) {
      throw new IssueParseError("Missing required field in issue data: title");
    }
    if (item.html_url === undefined) {
      throw new IssueParseError("Missing required field in issue data: html_url");
    }

    return {
      number: item.number,
      title: item.title,
      body: item.body ?? null,
      labels,
      url: item.html_url,
      authorAssociation: item.author_association ?? "",
      phase,
      priority,
    };
  });
}

/**
 * ラベルから優先度を判定する。
 *
 * priorityLabels のインデックスで判定:
 *   - index 0 のラベルが含まれる -> Priority.HIGH
 *   - index 1 のラベルが含まれる -> Priority.LOW
 *   - どちらも含まれない -> Priority.NONE
 *
 * 複数マッチ時は最も優先度が高いもの（インデックスが小さい方）を採用する。
 */
function determinePriority(
  labels: readonly string[],
  priorityLabels: readonly string[],
): Priority {
  const priorityMap: Record<number, Priority> = {
    0: Priority.HIGH,
    1: Priority.LOW,
  };

  for (let index = 0; index < priorityLabels.length; index++) {
    if (labels.includes(priorityLabels[index]) && index in priorityMap) {
      return priorityMap[index];
    }
  }

  return Priority.NONE;
}

/**
 * authorAssociation が許可リストに含まれない Issue を除外する。
 * 除外された Issue は WARNING ログに記録する。
 */
function filterByAuthorAssociation(issues: Issue[]): Issue[] {
  return issues.filter((issue) => {
    if (PERMITTED_ASSOCIATIONS.has(issue.authorAssociation)) {
      return true;
    }
    logger.warn(
      "Issue #%s skipped: author association '%s' is not permitted",
      issue.number,
      issue.authorAssociation,
    );
    return false;
  });
}

/**
 * Issue を優先度順にソートする。
 *
 * ソートキー: (priority, number) の昇順
 */
function sortByPriority(issues: Issue[]): Issue[] {
  return [...issues].sort((a, b) => {
    if (a.priority !== b.priority) {
      return a.priority - b.priority;
    }
    return a.number - b.number;
  });
}
